src/auth/pluggableauthclient.ts

```diff
@@ -0,0 +1,253 @@
+     */
+    timeout_millis?: number;
+    /**
+     * Optional output file location that will be checked for a cached response
+     * from a previous run of the executable.
+     */
+    output_file?: string;
+  };
+}
+
+/**
+ * Error class for errors thrown from executable run by PluggableAuthClient.
+ */
+export class ExecutableError extends Error {
+  /**
+   * Exit code returned by the executable.
+   */
+  readonly code: string;
+
+  constructor(message: string, code: string) {
+    super(message);
+    this.name = 'ExecutableError';
+    this.code = code;
+    Object.setPrototypeOf(this, ExecutableError.prototype);
+  }
+}
+
+/**
+ * Default executable timeout when none is provided, in milliseconds.
+ */
+const DEFAULT_EXECUTABLE_TIMEOUT_MILLIS = 30 * 1000;
+/**
+ * Minimum allowed executable timeout in milliseconds.
+ */
+const MINIMUM_EXECUTABLE_TIMEOUT_MILLIS = 5 * 1000;
+/**
+ * Maximum allowed executable timeout in milliseconds.
+ */
+const MAXIMUM_EXECUTABLE_TIMEOUT_MILLIS = 120 * 1000;
+
+/**
+ * Environment variable to check to see if executable can be run.
+ * Value must be set to '1' for the executable to run.
+ */
+const GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES =
+  'GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES';
+
+/**
+ * Maximum currently supported executable version.
+ */
+const MAXIMUM_EXECUTABLE_VERSION = 1;
+
+/**
+ * Pluggable auth external account client. This is used to call a user provided
+ * executable which returns a subject token to be exchanged for a
+ * Google access token.
+ */
+export class PluggableAuthClient extends BaseExternalAccountClient {
+  /**
+   * Command used to retrieve the third party token.
+   */
+  private readonly command: string;
+  /**
+   * Timeout in milliseconds for running executable,
+   * set to default if none provided.
+   */
+  private readonly timeoutMillis: number;
+  /**
+   * Path to file to check for cached executable response.
+   */
+  private readonly outputFile?: string;
+
+  /**
+   * Instantiates a PluggableAuthClient instance using the provided JSON
+   * Object loaded from an external account credentials file.
+   * An error is thrown if the credential is not a valid pluggable auth credential.
+   * @param options The external account options object typically loaded from
+   *   the external account JSON credential file.
+   * @param additionalOptions Optional additional behavior customization
+   *   options. These currently customize expiration threshold time and
+   *   whether to retry on 401/403 API request errors.
+   */
+  constructor(
+    options: PluggableAuthClientOptions,
+    additionalOptions?: RefreshOptions
+  ) {
+    super(options, additionalOptions);
+    this.command = options.credential_source.command;
+    if (!this.command) {
+      throw new Error('No valid Pluggable Auth "credential_source" provided.');
+    }
+    // Check if the provided timeout exists and if it is valid.
+    if (options.credential_source.timeout_millis === undefined) {
```

Comment:
!options.credential_source.timeout_millis?

Reply:
There's an edge case here I found in testing where if a timeout of 0 is provided, then that would return false and the default timeout would be used instead of throwing the invalid timeout error. I don't think its a huge deal either way since a user provided timeout of 0 doesn't make any sense in the first place, but I think having it be  explicit what we are checking here and only using the default if the field isn't provided makes sense.